clients/pkg/promtail/targets/file/decompresser.go

```diff
@@ -0,0 +1,285 @@
+package file
+
+import (
+	"bufio"
+	"compress/bzip2"
+	"compress/flate"
+	"compress/gzip"
+	"compress/zlib"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"go.uber.org/atomic"
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/encoding/ianaindex"
+	"golang.org/x/text/transform"
+
+	"github.com/grafana/loki/pkg/logproto"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/positions"
+)
+
+type decompresser struct {
+	metrics   *Metrics
+	logger    log.Logger
+	handler   api.EntryHandler
+	positions positions.Positions
+
+	path string
+
+	posAndSizeMtx sync.Mutex
+	stopOnce      sync.Once
+
+	running *atomic.Bool
+	posquit chan struct{}
+	posdone chan struct{}
+	done    chan struct{}
+
+	decoder *encoding.Decoder
+
+	position int64
+	size     int64
+}
+
+func newDecompresser(metrics *Metrics, logger log.Logger, handler api.EntryHandler, positions positions.Positions, path string, encodingFormat string) (*decompresser, error) {
+	logger = log.With(logger, "component", "decompresser")
+
+	pos, err := positions.Get(path)
+	if err != nil {
+		return nil, errors.Wrap(err, "get positions")
+	}
+
+	var decoder *encoding.Decoder
+	if encodingFormat != "" {
+		level.Info(logger).Log("msg", "decompresser will decode messages", "from", encodingFormat, "to", "UTF8")
+		encoder, err := ianaindex.IANA.Encoding(encodingFormat)
+		if err != nil {
+			return nil, errors.Wrap(err, "error doing IANA encoding")
+		}
+		decoder = encoder.NewDecoder()
+	}
+
+	decompresser := &decompresser{
+		metrics:   metrics,
+		logger:    logger,
+		handler:   api.AddLabelsMiddleware(model.LabelSet{FilenameLabel: model.LabelValue(path)}).Wrap(handler),
+		positions: positions,
+		path:      path,
+		running:   atomic.NewBool(false),
+		posquit:   make(chan struct{}),
+		posdone:   make(chan struct{}),
+		done:      make(chan struct{}),
+		position:  pos,
+		decoder:   decoder,
+	}
+
+	go decompresser.readLines()
+	go decompresser.updatePosition()
+	metrics.filesActive.Add(1.)
+	return decompresser, nil
+}
+
+// mountReader instantiate a reader ready to be used by the decompresser.
+//
+// The selected reader implementation is based on the extension of the given file name.
+// It'll error if the extension isn't supported.
+func mountReader(f *os.File, logger log.Logger) (reader io.Reader, err error) {
```

Comment:
I think `.tar` should be supported here as well - it's a very common format
Speaking of which, if a file is a compressed tarball (.tar,gz) that might need some special handling

Reply:
But tar is a container or archive format, but not a compression format. 
The feature of being able to read from compressed files should all about compressed data streams, not handling archive files with more files in them.

Reply:
Oh, it seems `.tar.gz` is supported:

```go
if strings.Contains(ext, "gz") { // .gz, .tar.gz
```

Sorry @DylanGuedes, I missed that

Reply:
np and no reason to apologize :smile: